### PR TITLE
Includes reorganization

### PR DESCRIPTION
**Pull request by valco1994**
* All source code files were rewritten to use conditional compilation on the base of headers/functions presence instead of platform-macro passing
* Also the problem with `pid_t` was solved (see issue #55) and bug in checking type sizes in `config.h.in` was fixed

### CODE REVIEW

**Comment by eyalroz** [comment]
1. Please separate the actual fixes from the include indentations. When everything is in perfect working order and no fixes are necessary, then we'll talk about formatting the includes... (and I'm still not convinced we should touch them at all.)
2. We should try to move away from  `#ifdef HAVE_FOO #include foo.h` pieces of code, as much as we can. Specifically, where you have an if-else-if-else-error sequence - that should go into `CMakeLists.txt` and trigger a failure during configuration.
3. We could use a trick similar to the one used [here](https://stackoverflow.com/questions/266501/macro-definition-containing-include-directive) to not have the if-have-include blocks in the C files. But I guess that's a bit too much for now.
2. We're not going to use GNU extensions, only standard C. If the person _building_ has themselves enabled GNU extensions, then we could be willing to work with that.

Bottom line: separate this PR out further... short and sweet PRs are what I like :-)

**Comment by valco1994** [comment]
1. Ok. I carried indentation out in a separate PR and thought, that it would be merged first. I'll delete that commit.
2. Several questions here: 
    * Should I do it now? This PR is about portability improvement, but you're talking about refactoring actually and doing it in this PR diverges with your idea "one PR - one thing". 
    * Such constructions (`#ifdef ... #include ... #endif`) can be carried out to `config.h.in`, but not all code files require them all, so it's not looking like a good solution. Or we can create a lot of config files (almost one per source code file). And IMHO it's looking even worse.
    * Talking about `#if-#else-#if-#else-#error` sequences: actually there is just one such sequence (in 2 places) - about getting `malloc()` from `stdlib.h` or `malloc.h`. And this check is made in CMakeLists.txt too with throwing FATAL_ERROR if a source of `malloc()` is not found. So, `#error` branch in the code was left just as double-checking, actually it should be unaccessible. I can delete it if you want.
3. I think, that it's not necessary for this PR, and it's looking overcomplicated for me in general.

**Comment by valco1994** [comment]
I got rid of indentation, watch, please

File: src/config.h.in [hunk 1]
```diff
@@ -27,9 +27,23 @@
...
+#cmakedefine HAVE_SYS_TYPES_H
+#cmakedefine HAVE_SYS_STAT_H
+#cmakedefine HAVE_STRINGS_H
 #cmakedefine HAVE_INTTYPES_H
+
+/*		Windows headers related macros		*/
+#cmakedefine HAVE_PROCESS_H
+#cmakedefine HAVE_WINDOWS_H	
+
+/*		Another headers related macros		*/
```
Comment (eyalroz) [inline src/config.h.in]: Can you change this to: `/* Non-platform-specific headers */`?

File: src/config.h.in [hunk 2]
```diff
@@ -27,9 +27,23 @@
...
+/*		POSIX headers related macros		*/
+#cmakedefine HAVE_UNISTD_H
+#cmakedefine HAVE_FCNTL_H
+#cmakedefine HAVE_SYS_WAIT_H
+#cmakedefine HAVE_SYS_TYPES_H
+#cmakedefine HAVE_SYS_STAT_H
+#cmakedefine HAVE_STRINGS_H
 #cmakedefine HAVE_INTTYPES_H
+
+/*		Windows headers related macros		*/
```
Comment (eyalroz) [inline src/config.h.in]: Can you make this just `/* Windows platform headers */`?

File: src/config.h.in [hunk 3]
```diff
@@ -27,9 +27,23 @@
  *   TPCH              -- make will create TPCH (set in makefile)
  */
 
+/*		POSIX headers related macros		*/
```
Comment (eyalroz) [inline src/config.h.in]: Can you make this just `/* POSIX platform headers */`?

File: src/config.h.in [hunk 4]
```diff
@@ -38,24 +52,18 @@
...
 #define SIGNED_64_BIT_C_TYPE long
 #define SIGNED_64_BIT_PRINTF_MODIFIER "l"
-#else
-#if (@SIZEOF_LONG_LONG@ == 8)
+#elif (@SIZEOF_LONG_LONG@ == 8)
 #define SIGNED_64_BIT_C_TYPE long long
 #define SIGNED_64_BIT_PRINTF_MODIFIER "ll"
-#else
-#if (@SIZEOF_SHORT@ == "8")
+#elif (@SIZEOF_SHORT@ == 8)
```
Comment (eyalroz) [inline src/config.h.in]: This is already resolved on the `additional_fixes` branch.

File: src/config.h.in [hunk 5]
```diff
@@ -38,24 +52,18 @@
...
 #define SIGNED_64_BIT_C_TYPE short
 #define SIGNED_64_BIT_PRINTF_MODIFIER "s"
-#else
-#if (@SIZEOF_INT@ == "8")
+#elif (@SIZEOF_INT@ == 8)
```
Comment (eyalroz) [inline src/config.h.in]: Ditto.

File: CMakeLists.txt [hunk 6]
```diff
@@ -68,9 +112,12 @@ add_executable(qgen
...
 
-# Link against the C math library (libm)
-target_link_libraries(dbgen m)
-target_link_libraries(qgen m)
+# Link with the C math library (libm) if required
+find_library(HAVE_LIBM m)
+if (HAVE_LIBM)
+	target_link_libraries(dbgen m)
+	target_link_libraries(qgen m)
+endif()
```
Comment (eyalroz) [inline CMakeLists.txt]: I've done something better on the `additional_fixes` branch, see #59 .

File: src/config.h.in [hunk 7]
```diff
@@ -74,38 +82,44 @@
...
+//#if (defined (__WATCOMC__))
+//#define SPAWN()   spawnv(P_NOWAIT, spawn_args[0], spawn_args)
+//#define WAIT(res, pid) cwait(res, pid, WAIT_CHILD)
+//#else /* MSVC (?) */
+//#define SPAWN()   _spawnv(_P_NOWAIT, spawn_args[0], spawn_args)
+//#define WAIT(res, pid) _cwait(res, pid, _WAIT_CHILD)
+//#define getpid          _getpid
+//#endif /* (defined (__WATCOMC__)) */
+//#endif /* (defined(HAVE_UNISTD_H) && defined(HAVE_SYS_WAIT_H)) */
+
```
Comment (eyalroz) [inline src/config.h.in]: I'm not super-happy about using `//` comments, since we don't have them anywhere else in the code. But... fine, I guess. Can you add `//-----` and `//----` delimiters and say "The following commented-out macros appeared in the original ssb-dbgen code, but were not being used, with parallel load only made possible for POSIX-compatible systems (using fork, kill and wait)."

File: src/driver.c [hunk 8]
```diff
@@ -6,26 +6,33 @@
...
 
 #include "config.h"
-#include <stdlib.h>
-#if ( defined(_POSIX_C_SOURCE) || !defined(WIN32) )		/* Change for Windows NT */
-#ifndef DOS
-#include <unistd.h>
-#include <sys/wait.h>
-#endif
+#if (defined(HAVE_FORK) && defined(HAVE_WAIT) && defined(HAVE_KILL))
+#define CAN_PARALLELIZE_DATA_GENERATION
```
Comment (eyalroz) [inline src/driver.c]: Great :+1:

**Comment by valco1994** [comment]
Ok, no problem
